### PR TITLE
Add `LightningCLI(auto_registry)`

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Adds a flag that will automatically register all subclasses in memory for the registerable components.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @carmocca @mauvilsa @tchaton

### CODE REVIEW

**Comment by carmocca** [comment]
I'd like to get this merged as is and then switch the default to `True` in a follow-up as that will require a larger docs rewrite and that way it's easier to revert if necessary.

File: pytorch_lightning/utilities/cli.py [hunk 1]
```diff
@@ -103,17 +104,29 @@ def __init__(self, optimizer: Optimizer, monitor: str, *args: Any, **kwargs: Any
...
-CALLBACK_REGISTRY = _Registry()
-CALLBACK_REGISTRY.register_classes(pl.callbacks, pl.callbacks.Callback)
-
-MODEL_REGISTRY = _Registry()
-
-DATAMODULE_REGISTRY = _Registry()
-
-LOGGER_REGISTRY = _Registry()
-LOGGER_REGISTRY.register_classes(pl.loggers, pl.loggers.LightningLoggerBase)
+def _fill_registries(subclasses: bool) -> None:
```
Comment (tchaton) [inline pytorch_lightning/utilities/cli.py:107]:
```suggestion
def _populate_registries(subclasses: bool) -> None:
```

File: pytorch_lightning/utilities/cli.py [hunk 2]
```diff
@@ -471,6 +484,7 @@ def __init__(
         subclass_mode_model: bool = False,
         subclass_mode_data: bool = False,
         run: bool = True,
+        register_automatically: bool = False,
```
Comment (tchaton) [inline pytorch_lightning/utilities/cli.py:487]:
```suggestion
        auto_register: bool = False,
```

**Review by tchaton — APPROVED** [review]
Love it ! And let's make this True by default.

**Review by SeanNaren — APPROVED** [review]

**Review by ethanwharris — APPROVED** [review]
LGTM 😃

**Pull request merged**